src/servers/src/http/jaeger.rs

```diff
@@ -0,0 +1,1244 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use axum::extract::{Path, Query, State};
+use axum::http::StatusCode;
+use axum::response::IntoResponse;
+use axum::Extension;
+use common_error::ext::ErrorExt;
+use common_telemetry::{debug, tracing, warn};
+use serde::{Deserialize, Serialize};
+use serde_json::Value as JsonValue;
+use session::context::{Channel, QueryContext, QueryContextRef};
+
+use crate::error::{FailedToExecuteJaegerQuerySnafu, InvalidJaegerQuerySnafu, Result};
+use crate::http::{handler, ApiState, GreptimeQueryOutput, HttpRecordsOutput};
+use crate::metrics::METRIC_JAEGER_QUERY_ELAPSED;
+use crate::otlp::trace::TRACE_TABLE_NAME;
+use crate::query_handler::sql::ServerSqlQueryHandlerRef;
+
+/// JaegerAPIResponse is the response of Jaeger HTTP API.
+/// The original version is `structuredResponse` which is defined in https://github.com/jaegertracing/jaeger/blob/main/cmd/query/app/http_handler.go.
+#[derive(Default, Debug, Serialize, Deserialize, PartialEq)]
+pub struct JaegerAPIResponse<T> {
+    pub data: Option<T>,
+    pub total: i32,
+    pub limit: i32,
+    pub offset: i32,
+    pub errors: Option<Vec<JaegerAPIError>>,
+}
+
+/// JaegerAPIError is the error of Jaeger HTTP API.
+#[derive(Default, Debug, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct JaegerAPIError {
+    pub code: i32,
+    pub msg: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub trace_id: Option<String>,
+}
+
+/// Operation is an operation in a service.
+#[derive(Debug, Default, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct Operation {
+    pub name: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub span_kind: Option<String>,
+}
+
+/// Trace is a collection of spans.
+#[derive(Debug, Default, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct Trace {
+    #[serde(rename = "traceID")]
+    pub trace_id: String,
+    pub spans: Vec<Span>,
+    pub processes: Option<HashMap<String, Process>>,
```

Comment:
The same suggestions as above for `processes` and `warnings`. Usually, we don't use `Option` to wrap vector or map; the empty means `None`.